File: web/lib/date-ranges.js
'use strict';

/**
 * web/lib/date-ranges.js
 *
 * The dashboard Summary panel's date-range presets. PURE — every function
 * takes the "current" date as an argument so the presets are testable and
 * so the server never depends on ambient clock state mid-request.
 *
 * Financial-year conventions (Indian FY): a year runs 1 Apr → 31 Mar, and
 * quarters are the FY quarters (Apr-Jun, Jul-Sep, Oct-Dec, Jan-Mar).
 * Weeks run Monday → Monday, matching the reference product.
 *
 * Labels render like "This Year (1st Apr '26 - 31st Mar '27)".
 */

const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
                'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

const RANGE_VALUES = [
    'today', 'yesterday', 'this_week', 'last_week',
    'this_month', 'last_month', 'this_quarter',
    'this_year', 'last_year',
];

const DEFAULT_RANGE = 'this_year';

// 1 → "1st", 2 → "2nd", 3 → "3rd", 11..13 → "th", else by last digit.
function ordinal(n) {
    const rem100 = n % 100;
    if (rem100 >= 11 && rem100 <= 13) return `${n}th`;
    switch (n % 10) {
        case 1:  return `${n}st`;
        case 2:  return `${n}nd`;
        case 3:  return `${n}rd`;
        default: return `${n}th`;
    }
}

// Date → 'YYYY-MM-DD' using LOCAL components (never toISOString, which would
// shift the date backwards for any positive UTC offset such as IST).
function iso(d) {
    const pad = (n) => String(n).padStart(2, '0');
    return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;
}

// Date → "3rd Aug '26".
function pretty(d) {
    const yy = String(d.getFullYear()).slice(-2);
    return `${ordinal(d.getDate())} ${MONTHS[d.getMonth()]} '${yy}`;
}

function addDays(d, n) {
    return new Date(d.getFullYear(), d.getMonth(), d.getDate() + n);
}

// Most recent Monday STRICTLY BEFORE d. Weeks are Monday→Monday inclusive of
// both ends (the reference product shows 27th Jul - 3rd Aug for a week whose
// "today" IS Monday 3rd Aug), so a Monday belongs to the week that opened the
// previous Monday rather than opening a zero-length week of its own.
function startOfWeek(d) {
    const dow = d.getDay();                 // 0 = Sunday
    const back = ((dow + 6) % 7) || 7;      // Tue → 1 … Sun → 6, Mon → 7
    return addDays(d, -back);
}

function startOfMonth(d, monthOffset = 0) {
    return new Date(d.getFullYear(), d.getMonth() + monthOffset, 1);
}

function endOfMonth(d, monthOffset = 0) {
    return new Date(d.getFullYear(), d.getMonth() + monthOffset + 1, 0);
}

// The calendar year in which this date's financial year STARTED.
// Jan/Feb/Mar belong to the FY that began the previous April.
function fyStartYear(d) {
    return d.getMonth() >= 3 ? d.getFullYear() : d.getFullYear() - 1;
}

// Financial quarter index 0..3 (Apr-Jun, Jul-Sep, Oct-Dec, Jan-Mar).
function fyQuarter(d) {
    return Math.floor(((d.getMonth() - 3 + 12) % 12) / 3);
}

function span(title, from, to) {
    return { title, from, to, label: `${title} (${pretty(from)} - ${pretty(to)})` };
}

function single(title, day) {
    return { title, from: day, to: day, label: `${title} (${pretty(day)})` };
}

/**
 * Build all nine presets for a given "now".
 * @returns {Array<{value:string,label:string,from:string,to:string}>}
 */
function buildRanges(now) {
    const d = now instanceof Date ? now : new Date();

    const weekStart = startOfWeek(d);
    const fyStart   = fyStartYear(d);
    const q         = fyQuarter(d);
    const qStart    = new Date(fyStart, 3 + q * 3, 1);
    const qEnd      = new Date(fyStart, 3 + q * 3 + 3, 0);

    const defs = {
        today:        single('Today', d),
        yesterday:    single('Yesterday', addDays(d, -1)),
        this_week:    span('This Week', weekStart, d),
        last_week:    span('Last Week', addDays(weekStart, -7), weekStart),
        this_month:   span('This Month', startOfMonth(d), endOfMonth(d)),
        last_month:   span('Last Month', startOfMonth(d, -1), endOfMonth(d, -1)),
        this_quarter: span('This Quarter', qStart, qEnd),
        this_year:    span('This Year', new Date(fyStart, 3, 1), new Date(fyStart + 1, 2, 31)),
        last_year:    span('Last Year', new Date(fyStart - 1, 3, 1), new Date(fyStart, 2, 31)),
    };

    return RANGE_VALUES.map((value) => {
        const def = defs[value];
        return { value, label: def.label, from: iso(def.from), to: iso(def.to) };
    });
}

/**
 * Resolve one preset by key. Unknown or missing keys fall back to this_year,
 * so a hand-edited URL can never produce an undefined range.
 */
function resolveRange(value, now) {
    const all = buildRanges(now);
    return all.find((r) => r.value === value) || all.find((r) => r.value === DEFAULT_RANGE);
}

module.exports = { RANGE_VALUES, DEFAULT_RANGE, buildRanges, resolveRange };

File: app/lib/core/utils/date_ranges.dart
/// core/utils/date_ranges.dart
///
/// Dart port of `web/lib/date-ranges.js` — the dashboard Summary panel's nine
/// date-range presets. PURE: every function takes "now" as an argument, so the
/// presets are testable and the app never depends on ambient clock state
/// mid-build.
///
/// Financial-year conventions (Indian FY): a year runs 1 Apr → 31 Mar, and
/// quarters are the FY quarters (Apr-Jun, Jul-Sep, Oct-Dec, Jan-Mar). Weeks
/// run Monday → Monday, matching the web.
///
/// Labels render like "This Year (1st Apr '26 - 31st Mar '27)" — identical
/// strings to the web's <select>, so the two dashboards read the same.
library;

const List<String> _months = [
  'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
  'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
];

/// One preset: the key sent to the API, its human label, and its ISO bounds.
class DateRange {
  const DateRange({
    required this.value,
    required this.label,
    required this.from,
    required this.to,
  });

  final String value;
  final String label;
  final String from; // YYYY-MM-DD
  final String to;   // YYYY-MM-DD
}

class DateRanges {
  DateRanges._();

  /// Preset keys, in the order the web's <select> lists them.
  static const List<String> values = [
    'today', 'yesterday', 'this_week', 'last_week',
    'this_month', 'last_month', 'this_quarter',
    'this_year', 'last_year',
  ];

  static const String defaultValue = 'this_year';

  // 1 → "1st", 2 → "2nd", 3 → "3rd", 11..13 → "th", else by last digit.
  static String _ordinal(int n) {
    final rem100 = n % 100;
    if (rem100 >= 11 && rem100 <= 13) return '${n}th';
    switch (n % 10) {
      case 1: return '${n}st';
      case 2: return '${n}nd';
      case 3: return '${n}rd';
      default: return '${n}th';
    }
  }

  /// DateTime → 'YYYY-MM-DD' from LOCAL components (never toIso8601String on a
  /// UTC value, which would shift the date back for a positive offset like IST).
  static String iso(DateTime d) {
    String p(int n) => n.toString().padLeft(2, '0');
    return '${d.year}-${p(d.month)}-${p(d.day)}';
  }

  /// DateTime → "3rd Aug '26".
  static String pretty(DateTime d) {
    final yy = d.year.toString().substring(2);
    return "${_ordinal(d.day)} ${_months[d.month - 1]} '$yy";
  }

  static DateTime _day(DateTime d) => DateTime(d.year, d.month, d.day);
  static DateTime _addDays(DateTime d, int n) => DateTime(d.year, d.month, d.day + n);

  /// Most recent Monday STRICTLY BEFORE [d] — a Monday belongs to the week that
  /// opened the *previous* Monday, exactly as the web computes it.
  static DateTime _startOfWeek(DateTime d) {
    final dow = d.weekday % 7;          // Dart: Mon=1..Sun=7 → JS: Sun=0..Sat=6
    final back = ((dow + 6) % 7) == 0 ? 7 : ((dow + 6) % 7);
    return _addDays(d, -back);
  }

  /// The calendar year in which this date's financial year STARTED.
  /// Jan/Feb/Mar belong to the FY that began the previous April.
  static int _fyStartYear(DateTime d) => d.month >= 4 ? d.year : d.year - 1;

  /// Financial quarter index 0..3 (Apr-Jun, Jul-Sep, Oct-Dec, Jan-Mar).
  static int _fyQuarter(DateTime d) => ((d.month - 4 + 12) % 12) ~/ 3;

  static DateRange _span(String value, String title, DateTime from, DateTime to) =>
      DateRange(
        value: value,
        label: '$title (${pretty(from)} - ${pretty(to)})',
        from: iso(from),
        to: iso(to),
      );

  static DateRange _single(String value, String title, DateTime day) => DateRange(
        value: value,
        label: '$title (${pretty(day)})',
        from: iso(day),
        to: iso(day),
      );

  /// All nine presets for a given "now", in [values] order.
  static List<DateRange> build(DateTime now) {
    final d = _day(now);
    final weekStart = _startOfWeek(d);
    final fyStart = _fyStartYear(d);
    final q = _fyQuarter(d);
    final qStart = DateTime(fyStart, 4 + q * 3, 1);
    final qEnd = DateTime(fyStart, 4 + q * 3 + 3, 0); // day 0 = last of prev month

    return [
      _single('today', 'Today', d),
      _single('yesterday', 'Yesterday', _addDays(d, -1)),
      _span('this_week', 'This Week', weekStart, d),
      _span('last_week', 'Last Week', _addDays(weekStart, -7), weekStart),
      _span('this_month', 'This Month',
          DateTime(d.year, d.month, 1), DateTime(d.year, d.month + 1, 0)),
      _span('last_month', 'Last Month',
          DateTime(d.year, d.month - 1, 1), DateTime(d.year, d.month, 0)),
      _span('this_quarter', 'This Quarter', qStart, qEnd),
      _span('this_year', 'This Year',
          DateTime(fyStart, 4, 1), DateTime(fyStart + 1, 3, 31)),
      _span('last_year', 'Last Year',
          DateTime(fyStart - 1, 4, 1), DateTime(fyStart, 3, 31)),
    ];
  }

  /// Resolve one preset by key. An unknown key falls back to [defaultValue],
  /// so a stale saved selection can never produce a null range.
  static DateRange resolve(String value, DateTime now) {
    final all = build(now);
    return all.firstWhere(
      (r) => r.value == value,
      orElse: () => all.firstWhere((r) => r.value == defaultValue),
    );
  }
}
